File: internal/trustlist/webauthn.go
package trustlist

import (
	"crypto/ecdsa"
	"crypto/ed25519"
	"crypto/elliptic"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
)

// authDataMinLen is the minimum WebAuthn authenticatorData length: 32 bytes
// rpIdHash + 1 byte flags + 4 bytes signCount = 37. We bounds-check before
// indexing any of those fields.
const authDataMinLen = 37

// flagUserPresent is the User-Present (UP) bit in authenticatorData flags. We
// require it. We deliberately do NOT gate on the signature counter
// (authData[33:37]) because synced passkeys legitimately emit a counter of 0.
const flagUserPresent = 0x01

// flagUserVerified is the User-Verified (UV) bit. We require it too: both ceremonies pin
// userVerification:"required" client-side, so the SERVER — the sole enforcement authority — must
// enforce UV. A User-Present-only assertion otherwise degrades the declared "user-verified" factor
// to mere possession. Unlike the counter, a UV-capable authenticator DOES set this under UV=required.
const flagUserVerified = 0x04

// verifyWebAuthn verifies a WebAuthn (FIDO2) assertion over a trust list, stdlib
// only, and binds the assertion to THIS trust list's CONTENT: the expected challenge
// is Challenge(tl) = SHA256(Canonical(tl)). It is a thin content-bound wrapper over
// verifyAssertion (the challenge-agnostic core); the operator passkey LOGIN path
// reuses that same core through VerifyAssertion with a server-issued RANDOM nonce.
func verifyWebAuthn(tl TrustList, art SignedTrustList, pin PinnedCredential) error {
	// CONTENT BINDING: the challenge the assertion must carry is base64url(Challenge(tl)),
	// proving the user authorized these exact trust-list bytes (not a random nonce).
	chal, err := Challenge(tl)
	if err != nil {
		return err
	}
	return verifyAssertion(art, pin, chal)
}

// VerifyAssertion verifies a WebAuthn (FIDO2) assertion against a pinned credential and
// an EXPECTED challenge — the RAW challenge bytes; clientDataJSON.challenge must equal
// their base64url encoding. It is the challenge-agnostic sibling of the keystone path,
// exported for callers whose challenge is NOT content-bound: operator passkey LOGIN,
// where the challenge is a server-issued, single-use RANDOM nonce rather than the
// manifest hash. The structural, relying-party, user-presence, and signature checks are
// byte-for-byte identical to the keystone verifier; ONLY the source of the expected
// challenge differs.
//
// Like Verify, dispatch is on pin.Alg and an art.Alg != pin.Alg is rejected
// (ErrAlgMismatch) before any cryptography. Only the WebAuthn algorithms are accepted: a
// raw (non-WebAuthn) Ed25519 pin has no assertion to verify and is rejected as
// unsupported here.
func VerifyAssertion(art SignedTrustList, pin PinnedCredential, challenge []byte) error {
	if art.Alg != pin.Alg {
		return fmt.Errorf("%w: pinned=%q artifact=%q", ErrAlgMismatch, pin.Alg, art.Alg)
	}
	switch pin.Alg {
	case AlgWebAuthnES256, AlgWebAuthnEdDSA:
		return verifyAssertion(art, pin, challenge)
	default:
		return fmt.Errorf("%w: %q (passkey login requires a WebAuthn credential)", ErrUnsupportedAlg, pin.Alg)
	}
}

// AssertionChallenge extracts the base64url challenge string embedded in a WebAuthn
// assertion's clientDataJSON. A caller that issued a RANDOM server-side challenge (e.g.
// operator passkey login) uses it as the lookup key for the single-use challenge record
// it stored, then passes the DECODED bytes to VerifyAssertion. The keystone path does
// NOT need this — its expected challenge is recomputed from the manifest, never looked
// up. It does no signature work; it only parses the (still-to-be-verified) client data.
func AssertionChallenge(art SignedTrustList) (string, error) {
	cData, err := base64.RawURLEncoding.DecodeString(art.ClientDataJSON)
	if err != nil {
		return "", fmt.Errorf("trustlist: decode client_data_json: %w", err)
	}
	var cd struct {
		Challenge string `json:"challenge"`
	}
	if err := json.Unmarshal(cData, &cd); err != nil {
		return "", fmt.Errorf("trustlist: parse client_data_json: %w", err)
	}
	return cd.Challenge, nil
}

// verifyAssertion is the challenge-agnostic WebAuthn assertion verifier, stdlib only.
// The caller supplies the EXPECTED challenge bytes (wantChallenge); everything else is
// shared between the content-bound keystone path (wantChallenge = Challenge(tl)) and the
// random-nonce login path (wantChallenge = the issued nonce).
//
// Security-critical sequence (every failure is fail-closed):
//  1. Decode authenticatorData and clientDataJSON from base64url. Note: the
//     EXACT received clientDataJSON bytes are hashed in step 8 — we never
//     re-marshal the parsed struct, because re-marshaling could change bytes and
//     break the signature binding (and would let a forger smuggle extra fields).
//  2. Bounds-check authenticatorData length (>= 37) before indexing.
//  3. Parse clientDataJSON for type / challenge / origin.
//  4. type must be "webauthn.get" (an assertion, not a registration).
//  5. challenge must equal base64url(wantChallenge) — the binding to the exact bytes
//     the verifier expected (content hash for the keystone, random nonce for login).
//  6. rpIdHash (authData[0:32]) must equal sha256(pin.RPID).
//  7. User-Present flag must be set.
//  8. signedMessage = authenticatorData || sha256(clientDataJSON).
//  9. Decode the signature from base64url.
//  10. Verify per pinned alg: ES256 -> ecdsa.VerifyASN1 over sha256(signed);
//     EdDSA -> ed25519.Verify over signed directly (Ed25519 signs the message,
//     not a pre-hash). Anything else (RS256, etc.) is rejected.
//
// pin.Origin, when non-empty, is checked against clientDataJSON.origin as an
// ADVISORY measure. It is not authoritative on a node (a node cannot prove which
// browser origin a user used months earlier), so a mismatch is reported but the
// origin check is documented as advisory; the content binding (step 5) is the
// real authority.
func verifyAssertion(art SignedTrustList, pin PinnedCredential, wantChallenge []byte) error {
	// 0. Pin config precondition: an empty RPID makes the rpIdHash check meaningless
	//    (sha256("") is a known constant), silently disabling relying-party binding.
	//    A keystone anchor must have a real RPID — reject fail-closed.
	if pin.RPID == "" {
		return fmt.Errorf("trustlist: webauthn pin has empty RPID (relying-party binding disabled)")
	}

	// 1. Decode the assertion components.
	authData, err := base64.RawURLEncoding.DecodeString(art.AuthenticatorData)
	if err != nil {
		return fmt.Errorf("trustlist: decode authenticator_data: %w", err)
	}
	cData, err := base64.RawURLEncoding.DecodeString(art.ClientDataJSON)
	if err != nil {
		return fmt.Errorf("trustlist: decode client_data_json: %w", err)
	}

	// 2. Bounds-check before indexing rpIdHash / flags.
	if len(authData) < authDataMinLen {
		return fmt.Errorf("trustlist: authenticator_data too short: %d bytes, want >= %d", len(authData), authDataMinLen)
	}

	// 3. Parse the client data. We only need these three fields; extra fields are
	//    ignored for parsing but remain part of the hashed bytes (step 8).
	var cd struct {
		Type      string `json:"type"`
		Challenge string `json:"challenge"`
		Origin    string `json:"origin"`
	}
	if err := json.Unmarshal(cData, &cd); err != nil {
		return fmt.Errorf("trustlist: parse client_data_json: %w", err)
	}

	// 4. Must be an assertion ("get"), not a registration ("create").
	if cd.Type != "webauthn.get" {
		return fmt.Errorf("trustlist: client_data type = %q, want \"webauthn.get\"", cd.Type)
	}

	// 5. CHALLENGE BINDING: challenge must be base64url(wantChallenge).
	want := base64.RawURLEncoding.EncodeToString(wantChallenge)
	if cd.Challenge != want {
		return ErrChallengeMismatch
	}

	// 6. rpIdHash must equal sha256(pin.RPID).
	wantRP := sha256.Sum256([]byte(pin.RPID))
	gotRP := authData[0:32]
	if !constEqual(gotRP, wantRP[:]) {
		return fmt.Errorf("trustlist: rpIdHash mismatch for RPID %q", pin.RPID)
	}

	// 7. User-Present flag must be set. (Counter authData[33:37] is intentionally
	//    NOT checked — synced passkeys emit 0.)
	if authData[32]&flagUserPresent == 0 {
		return fmt.Errorf("trustlist: User-Present flag not set in authenticator_data")
	}

	// 7b. User-Verified flag must ALSO be set — the ONE gate shared by operator login, 2FA, and
	//     keystone off-host signing (webauthn.ts runAssertion pins userVerification:"required" for
	//     all). The server is the sole enforcement authority; a User-Present-ONLY assertion (a PIN-less
	//     authenticator, or a tampered client requesting UV=discouraged) would otherwise mint a full
	//     operator session / sign a manifest with mere "touch", collapsing the "user-verified" factor
	//     to possession. A UV-capable authenticator sets this under UV=required, so honest ceremonies
	//     are unaffected. (This gate ALSO runs node-side in VerifyMembership — the operator's manifest
	//     signature must carry UV — so a live fleet's existing manifests must have been UV-signed.)
	if authData[32]&flagUserVerified == 0 {
		return ErrUserVerification
	}

	// Advisory origin check (not authoritative on a node).
	if pin.Origin != "" && cd.Origin != pin.Origin {
		return fmt.Errorf("trustlist: client_data origin = %q, want %q (advisory)", cd.Origin, pin.Origin)
	}

	// 8. signedMessage = authenticatorData || SHA-256(clientDataJSON). Use the
	//    EXACT received cData bytes, never a re-marshal.
	cHash := sha256.Sum256(cData)
	signed := make([]byte, 0, len(authData)+len(cHash))
	signed = append(signed, authData...)
	signed = append(signed, cHash[:]...)

	// 9. Decode the assertion signature.
	sig, err := base64.RawURLEncoding.DecodeString(art.Signature)
	if err != nil {
		return fmt.Errorf("trustlist: decode webauthn signature: %w", err)
	}

	// 10. Verify per the PINNED algorithm.
	switch pin.Alg {
	case AlgWebAuthnES256:
		// Re-assert the pin shape at the verify site (defense in depth): a nil key or
		// nil/non-P-256 curve would panic inside ecdsa.VerifyASN1. ParseES256Pin already
		// enforces P-256, but a pin constructed by other means must still fail closed.
		if pin.ES256Pub == nil || pin.ES256Pub.Curve != elliptic.P256() {
			return ErrMissingPin
		}
		// ES256 = ECDSA P-256 with SHA-256; the signature is DER (ASN.1) over
		// SHA-256(signedMessage).
		digest := sha256.Sum256(signed)
		if !ecdsa.VerifyASN1(pin.ES256Pub, digest[:], sig) {
			return ErrBadSignature
		}
		return nil
	case AlgWebAuthnEdDSA:
		// Require the exact key size: ed25519.Verify PANICS on a wrong-length public
		// key (vs returning false), so a malformed pin must fail closed here, not crash.
		if len(pin.Ed25519Pub) != ed25519.PublicKeySize {
			return ErrMissingPin
		}
		// Ed25519 signs the message directly (it hashes internally); do NOT
		// pre-hash signedMessage.
		if !ed25519.Verify(pin.Ed25519Pub, signed, sig) {
			return ErrBadSignature
		}
		return nil
	default:
		// RS256 or any other algorithm is explicitly excluded.
		return fmt.Errorf("%w: webauthn alg %q", ErrUnsupportedAlg, pin.Alg)
	}
}

// constEqual is a constant-time byte-slice comparison wrapper kept local to
// avoid pulling crypto/subtle into the public surface; both inputs here are
// non-secret hashes, but constant-time is a harmless good habit.
func constEqual(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	var v byte
	for i := range a {
		v |= a[i] ^ b[i]
	}
	return v == 0
}

File: frontend/src/lib/webauthn.ts
// WebAuthn (FIDO2) ceremony helpers shared by login passkeys and the keystone operator
// signing key. New credentials are created, then immediately make a UV-requesting assertion
// over a controller nonce before either public descriptor is persisted. Existing credentials
// continue to drive login and content-bound manifest assertions through the same wire shape.
//
// The bytes produced here must verify byte-for-byte against the Go node verifier
// in internal/trustlist/webauthn.go. The exact wire contract is:
//
//   SignedTrustList (POST /trustlist-signature `signed` field):
//     { alg, credential_id, public_key, signature,
//       authenticator_data, client_data_json }
//   — alg          : "webauthn-es256" | "webauthn-eddsa"
//   — credential_id: base64url(rawId)                       (RawURLEncoding)
//   — public_key   : the pinned PKIX PEM (audit only, never trusted by node)
//   — signature    : base64url(response.signature)          (RawURLEncoding)
//                    ES256 returns ASN.1 DER (ecdsa.VerifyASN1 on the Go side)
//   — authenticator_data: base64url(response.authenticatorData)
//   — client_data_json  : base64url(response.clientDataJSON)
//
//   Browser credential enrollment (POST /passkey/register or /operator-credential):
//     { alg, credential_id, public_key_pem, rpid, origin, enrollment_proof }
//   — public_key_pem: PKIX "PUBLIC KEY" PEM wrapping the SPKI DER returned by
//                     cred.response.getPublicKey()
//   — rpid          : location.hostname (node checks SHA256(rpid)==authData
//                     rpIdHash, so rp.id at create() MUST equal this)
//   — origin        : location.origin
//   — enrollment_proof: a SignedTrustList-shaped assertion over the one-use server challenge
//
// CHALLENGE BINDING: the node checks clientData.challenge ==
// base64url(SHA256(Canonical(manifest))). The browser base64url-encodes the
// challenge ArrayBuffer into clientDataJSON automatically, so we pass the RAW
// SHA-256 digest of the DECODED canonical manifest bytes as the challenge — we
// do NOT pre-encode it.
//
// All base64 / base64url conversions and PEM wrapping are done by hand so this
// module pulls in no new npm dependency.

import { AlgWebAuthnES256, AlgWebAuthnEdDSA } from '../api/controllerClient';
import type { WebAuthnAlg, SignedTrustList } from '../api/controllerClient';

// COSE algorithm identifiers returned by getPublicKeyAlgorithm(). We accept only
// ES256 (-7) and EdDSA (-8); anything else (notably RS256 = -257) is rejected so
// the pinned credential always matches one of the two algorithms the node
// verifier dispatches on.
//
// ES256 is the PRIMARY, production path: it is universally supported by passkeys
// and YubiKeys, and is the one to trust. EdDSA is best-effort — OKP/Ed25519
// getPublicKey() support varies by platform/browser, and the live ceremony
// cannot be exercised in CI (no authenticator) — but both verify against the
// node when present, and the authenticator picks the first it supports.
const COSE_ES256 = -7;
const COSE_EDDSA = -8;

// WebAuthnError is the typed error this module throws. `kind` lets the UI react
// differently to a user cancel/timeout (NotAllowedError) vs. an unsupported
// platform vs. a rejected algorithm — without string-matching messages.
export type WebAuthnErrorKind =
  | 'unsupported' // navigator.credentials / WebAuthn not available
  | 'cancelled' // user dismissed the prompt or it timed out (NotAllowedError)
  | 'unsupported-algorithm' // authenticator returned a non-ES256/EdDSA key
  | 'no-public-key' // getPublicKey() returned null (no SPKI available)
  | 'invalid-rp-id' // RP ID is an IP literal (e.g. panel opened at http://127.0.0.1)
  | 'enrollment-verification-failed' // create succeeded, but candidate proof/persistence did not
  | 'failed'; // any other ceremony failure

export class WebAuthnError extends Error {
  readonly kind: WebAuthnErrorKind;
  constructor(kind: WebAuthnErrorKind, message: string) {
    super(message);
    this.name = 'WebAuthnError';
    this.kind = kind;
  }
}

// WebAuthn forbids IP-address RP IDs: per spec the RP ID must be a registrable
// domain, and 'localhost' is the only non-domain browsers special-case. When the
// panel is opened at http://127.0.0.1:PORT (or [::1]), location.hostname is an IP
// literal, so navigator.credentials.create()/.get() reject it with an opaque
// "This is an invalid domain." Catch it up front and tell the operator exactly how
// to fix it — browse to http://localhost:PORT, which resolves to the same loopback
// container but presents a valid RP ID.
// `host` is always a window.location.hostname value, which the URL parser has
// already normalized — IPv4 shorthand (127.1, 2130706433) is expanded to a
// dotted-quad before it reaches us, so the strict 4-octet regex still catches it.
function isIpLiteralHost(host: string): boolean {
  const h = host.replace(/^\[|\]$/g, ''); // strip IPv6 brackets if location kept them
  if (/^\d{1,3}(\.\d{1,3}){3}$/.test(h)) return true; // IPv4 dotted-quad
  if (h.includes(':')) return true; // IPv6 — a colon never appears in a DNS hostname
  return false;
}

function assertRegistrableRpId(rpId: string): void {
  if (!isIpLiteralHost(rpId)) return;
  const port =
    typeof location !== 'undefined' && location.port ? `:${location.port}` : '';
  throw new WebAuthnError(
    'invalid-rp-id',
    `WebAuthn can't use the IP address "${rpId}" as its domain. Open the panel at ` +
      `http://localhost${port} (not an IP address like 127.0.0.1 or ::1), or use a ` +
      `real hostname behind a reverse proxy, then register the passkey again.`,
  );
}

// Public descriptor extracted after navigator.credentials.create(). It is a candidate—not yet
// enrolled server state—and is safe to keep in volatile UI memory so a failed second phase can
// retry the same credential instead of creating duplicates in the authenticator.
export interface WebAuthnCredentialCandidate {
  alg: WebAuthnAlg;
  credentialId: string; // base64url(rawId)
  publicKeyPEM: string; // PKIX "PUBLIC KEY" PEM
  // Creation-context binding. These public values are part of a WebAuthn credential's
  // verification identity, so lost-response reconciliation must compare them alongside the
  // key/algorithm/id rather than treating a same-key credential from another RP/origin as exact.
  rpId: string;
  origin: string;
}

// --- base64 / base64url helpers (hand-rolled, no deps) ---

// base64url-encode bytes with the RFC4648 url alphabet and NO padding — matches
// Go's base64.RawURLEncoding, which every base64url field on the wire uses.
function bytesToBase64Url(bytes: Uint8Array): string {
  let bin = '';
  for (let i = 0; i < bytes.length; i++) {
    bin += String.fromCharCode(bytes[i]);
  }
  return btoa(bin).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

// Standard base64 (with padding) of bytes — used only for the PKIX PEM body,
// which is standard base64 per RFC 7468.
function bytesToBase64Std(bytes: Uint8Array): string {
  let bin = '';
  for (let i = 0; i < bytes.length; i++) {
    bin += String.fromCharCode(bytes[i]);
  }
  return btoa(bin);
}

// Wrap SPKI DER bytes into a PKIX "PUBLIC KEY" PEM block with 64-char lines, the
// shape ParseES256Pin / ParseEd25519PinPEM expect on the Go side.
function spkiToPEM(spki: Uint8Array): string {
  const b64 = bytesToBase64Std(spki);
  const lines: string[] = [];
  for (let i = 0; i < b64.length; i += 64) {
    lines.push(b64.slice(i, i + 64));
  }
  return `-----BEGIN PUBLIC KEY-----\n${lines.join('\n')}\n-----END PUBLIC KEY-----\n`;
}

// --- random user id (enrollment only) ---

// A fresh random n-byte buffer, explicitly backed by an ArrayBuffer (not a
// SharedArrayBuffer) so it satisfies the BufferSource the WebAuthn options want.
// Used for the opaque random user.id handle. Enrollment challenges come from the
// controller and are consumed once when the post-create UV proof is submitted.
function randomBytes(n: number): Uint8Array<ArrayBuffer> {
  const b = new Uint8Array(new ArrayBuffer(n));
  crypto.getRandomValues(b);
  return b;
}

// Map a COSE algorithm identifier to our wire Alg, rejecting anything but
// ES256 / EdDSA (RS256 and friends are excluded fail-closed).
function algFromCOSE(cose: number): WebAuthnAlg {
  switch (cose) {
    case COSE_ES256:
      return AlgWebAuthnES256;
    case COSE_EDDSA:
      return AlgWebAuthnEdDSA;
    default:
      throw new WebAuthnError(
        'unsupported-algorithm',
        `authenticator returned COSE algorithm ${cose}; only ES256 (-7) and EdDSA (-8) are supported`,
      );
  }
}

// Guard: WebAuthn must be available in this context (secure origin + a browser
// that exposes PublicKeyCredential).
function assertWebAuthnAvailable(): void {
  if (
    typeof navigator === 'undefined' ||
    !navigator.credentials ||
    typeof PublicKeyCredential === 'undefined'
  ) {
    throw new WebAuthnError(
      'unsupported',
      'WebAuthn is not available in this browser (a secure https/localhost context is required)',
    );
  }
}

// Normalize a thrown DOMException/Error into a typed WebAuthnError. A
// NotAllowedError is the browser's signal for a user cancel OR a timeout — we
// surface it as 'cancelled' so the UI can prompt a retry rather than show a
// scary failure.
function toWebAuthnError(err: unknown, fallback: string): WebAuthnError {
  if (err instanceof WebAuthnError) {
    return err;
  }
  if (err instanceof DOMException && err.name === 'NotAllowedError') {
    return new WebAuthnError('cancelled', 'the security-key prompt was dismissed or timed out');
  }
  const msg = err instanceof Error ? err.message : String(err);
  return new WebAuthnError('failed', `${fallback}: ${msg}`);
}

// --- enrollment: create a login or off-host operator credential ---

// createWebAuthnCredentialCandidate drives navigator.credentials.create() to mint a new login
// or off-host signing credential, then extracts its SPKI public key + COSE alg via
// the MODERN WebAuthn API (getPublicKey / getPublicKeyAlgorithm), avoiding CBOR
// attestation parsing entirely.
//
// rpId MUST equal the rpid you POST to /operator-credential (the node verifier
// checks SHA256(rpid)==authData rpIdHash), so callers pass location.hostname.
// origin is recorded for the advisory origin check on the node.
export async function createWebAuthnCredentialCandidate(
  rpId: string,
  origin: string,
  enrollmentChallenge: string,
): Promise<WebAuthnCredentialCandidate> {
  assertWebAuthnAvailable();
  // Reject an IP-literal RP ID with actionable guidance before the browser throws
  // its opaque "invalid domain" (panel opened at http://127.0.0.1 instead of localhost).
  assertRegistrableRpId(rpId);

  // Sanity-check the caller's recorded origin against the live browser origin: the
  // credential is created in THIS context, and that same origin is what the node's
  // (advisory) origin check compares clientData.origin against later. A mismatch
  // means the caller is about to PIN an origin the browser won't actually present,
  // which would trip the node's advisory check on every signed deploy.
  if (typeof location !== 'undefined' && origin !== location.origin) {
    throw new WebAuthnError(
      'failed',
      `origin mismatch: pinning ${origin} but this context is ${location.origin}`,
    );
  }

  const options: PublicKeyCredentialCreationOptions = {
    // Server-issued, one-use challenge. After create(), the new credential immediately
    // asserts over the same nonce; the controller consumes and verifies that UV proof
    // before accepting the public key.
    challenge: base64UrlToBytes(enrollmentChallenge),
    rp: {
      // id MUST equal the rpid posted to the controller; the node binds
      // SHA256(rpid) against the assertion's rpIdHash at verify time.
      id: rpId,
      name: 'YAOG',
    },
    user: {
      id: randomBytes(16),
      name: 'YAOG operator',
      displayName: 'YAOG operator',
    },
    // ES256 first, EdDSA second — ES256 + EdDSA ONLY (no RS256). The node
    // verifier dispatches only on these two; the authenticator picks the first
    // it supports.
    pubKeyCredParams: [
      { type: 'public-key', alg: COSE_ES256 },
      { type: 'public-key', alg: COSE_EDDSA },
    ],
    authenticatorSelection: {
      userVerification: 'required',
    },
    timeout: 120000,
    attestation: 'none',
  };

  let cred: PublicKeyCredential | null;
  try {
    cred = (await navigator.credentials.create({
      publicKey: options,
    })) as PublicKeyCredential | null;
  } catch (err) {
    throw toWebAuthnError(err, 'failed to enroll WebAuthn credential');
  }
  if (!cred) {
    throw new WebAuthnError('failed', 'no credential was created');
  }

  const response = cred.response as AuthenticatorAttestationResponse;
  if (typeof response.getPublicKey !== 'function') {
    throw new WebAuthnError(
      'unsupported',
      'this browser does not expose getPublicKey() on the attestation response',
    );
  }

  const alg = algFromCOSE(response.getPublicKeyAlgorithm());
  const spki = response.getPublicKey();
  if (!spki) {
    throw new WebAuthnError('no-public-key', 'the authenticator did not return a public key');
  }

  const credentialId = bytesToBase64Url(new Uint8Array(cred.rawId));
  const publicKeyPEM = spkiToPEM(new Uint8Array(spki));

  return { alg, credentialId, publicKeyPEM, rpId, origin };
}

// proveWebAuthnCredentialEnrollment asks the exact candidate to assert over the controller's
// one-use nonce with userVerification:"required". The controller verifies the signature and
// signed UV result bit under the submitted candidate key before persisting it. YAOG requests no
// attestation, so this validates the first-party browser ceremony; it does not establish hardware
// provenance, non-exportability, or resistance to a custom client submitting a software key.
export async function proveWebAuthnCredentialEnrollment(
  candidate: WebAuthnCredentialCandidate,
  rpId: string,
  enrollmentChallenge: string,
): Promise<SignedTrustList> {
  try {
    return await runAssertion(
      base64UrlToBytes(enrollmentChallenge),
      candidate.credentialId,
      candidate.alg,
      rpId,
      candidate.publicKeyPEM,
      'required',
      'failed to complete PIN/biometric verification for the new credential',
    );
  } catch (err) {
    const detail = toWebAuthnError(
      err,
      'failed to complete PIN/biometric verification for the new credential',
    );
    throw new WebAuthnError(
      'enrollment-verification-failed',
      `The credential was created, but YAOG did not enroll it because PIN/biometric verification did not complete. ` +
        `YAOG kept this candidate for the current session; retry enrollment verification instead of creating another credential. (${detail.message})`,
    );
  }
}

// --- signing: produce a content-bound SignedTrustList over a manifest ---

// signManifest drives navigator.credentials.get() to sign the EXACT canonical
// manifest bytes the controller staged, returning a SignedTrustList ready for
// POST /trustlist-signature.
//
// CONTENT BINDING: challenge = SHA-256(manifestBytes). manifestBytes are the
// DECODED canonical bytes (the caller base64-decodes the standard-base64
// trustlist_json from GET /trustlist before calling this). The browser
// base64url-encodes this digest into clientDataJSON.challenge, which the node
// compares against base64url(SHA256(Canonical(manifest))) — so this is the
// proof the operator authorized THESE exact bytes.
//
// rpId MUST equal the rpid that was pinned at enrollment (the node checks
// SHA256(rpid)==authData rpIdHash). We pass it EXPLICITLY rather than letting the
// browser default rpId to the caller's effective domain: the implicit default
// only happens to match when enroll and sign run from the same hostname, and a
// mismatch would surface as an opaque node-side rpIdHash failure (a deploy-time
// 400) instead of a clear browser error. Pinning it makes the binding explicit.
//
// publicKeyPEM is the pinned PEM (audit-only field on the wire); pass the value
// stored at enrollment.
export async function signManifest(
  manifestBytes: Uint8Array,
  credentialId: string,
  alg: WebAuthnAlg,
  rpId: string,
  publicKeyPEM: string,
): Promise<SignedTrustList> {
  assertWebAuthnAvailable();

  // challenge = raw SHA-256 of the decoded canonical manifest bytes. Pass it
  // RAW — the browser base64url-encodes it into clientDataJSON for us; do NOT
  // pre-encode it (that would double-encode and break the content binding).
  const digest = await crypto.subtle.digest('SHA-256', manifestBytes as BufferSource);
  return runAssertion(
    new Uint8Array(digest),
    credentialId,
    alg,
    rpId,
    publicKeyPEM,
    'preferred',
    'failed to sign the deploy manifest',
  );
}

// --- login: produce a random-challenge SignedTrustList (operator passkey login) ---

// assertLogin drives navigator.credentials.get() to prove possession of a registered
// LOGIN passkey over a server-issued RANDOM challenge (the sibling of signManifest's
// content-bound manifest hash). It is used by both the password+passkey 2FA leg and the
// passwordless flow. challengeB64 is the base64url nonce from the controller (passkey_
// required / begin); credentialId + alg + rpId come from the same response. The returned
// SignedTrustList's public_key is empty: the controller verifies against the PINNED
// credential it stored, never this audit-only field.
export async function assertLogin(
  challengeB64: string,
  credentialId: string,
  alg: WebAuthnAlg,
  rpId: string,
): Promise<SignedTrustList> {
  // The login challenge is a RANDOM nonce, not content-bound — decode it to the raw
  // bytes the browser signs over (do NOT hash it, unlike signManifest).
  return runAssertion(
    base64UrlToBytes(challengeB64),
    credentialId,
    alg,
    rpId,
    '',
    'preferred',
    'failed to complete the passkey login',
  );
}

// runAssertion is the shared navigator.credentials.get() ceremony: it asserts the pinned
// credential over `challenge` (raw bytes the browser base64url-encodes into
// clientDataJSON) and assembles the SignedTrustList wire struct. Both signManifest
// (challenge = manifest hash), assertLogin (challenge = random nonce), and the enrollment proof
// (challenge = purpose-scoped random nonce) delegate here. Enrollment requires UV; ordinary
// assertions prefer it without excluding historical non-UV credentials.
async function runAssertion(
  challenge: Uint8Array,
  credentialId: string,
  alg: WebAuthnAlg,
  rpId: string,
  publicKeyPEM: string,
  userVerification: UserVerificationRequirement,
  failMessage: string,
): Promise<SignedTrustList> {
  assertWebAuthnAvailable();
  // Same IP-literal guard as enrollment: .get() rejects an IP RP ID identically, so
  // surface the actionable "use localhost" message instead of the opaque browser error.
  assertRegistrableRpId(rpId);

  // Restrict the assertion to the pinned credential so the browser uses exactly the key
  // the controller has on file. Pin rpId explicitly (do not rely on the effective-domain
  // default), matching the rpid the node binds SHA256(rpid)==rpIdHash against.
  const options: PublicKeyCredentialRequestOptions = {
    challenge: challenge as BufferSource,
    rpId,
    allowCredentials: [{ type: 'public-key', id: base64UrlToBytes(credentialId) as BufferSource }],
    // Only enrollment passes "required". Ordinary login and signing use "preferred" so a
    // credential that worked before rc.7 is never rejected merely because it cannot perform UV.
    userVerification,
    timeout: 120000,
  };

  let cred: PublicKeyCredential | null;
  try {
    cred = (await navigator.credentials.get({ publicKey: options })) as PublicKeyCredential | null;
  } catch (err) {
    throw toWebAuthnError(err, failMessage);
  }
  if (!cred) {
    throw new WebAuthnError('failed', 'no assertion was returned');
  }

  const response = cred.response as AuthenticatorAssertionResponse;
  return {
    alg,
    credential_id: bytesToBase64Url(new Uint8Array(cred.rawId)),
    // Audit-only on the wire; the node/controller verifies against the PINNED key, never
    // this field. signManifest echoes the enrolled PEM; login leaves it empty.
    public_key: publicKeyPEM,
    // ES256 assertions are ASN.1 DER (ecdsa.VerifyASN1 on the Go side); pass
    // response.signature through verbatim, base64url-encoded.
    signature: bytesToBase64Url(new Uint8Array(response.signature)),
    authenticator_data: bytesToBase64Url(new Uint8Array(response.authenticatorData)),
    client_data_json: bytesToBase64Url(new Uint8Array(response.clientDataJSON)),
  };
}

// base64url (no padding) -> bytes. Mirrors Go's base64.RawURLEncoding.Decode;
// used to turn the stored credential_id back into an allowCredentials id.
function base64UrlToBytes(s: string): Uint8Array<ArrayBuffer> {
  const b64 = s.replace(/-/g, '+').replace(/_/g, '/');
  const bin = atob(b64);
  const out = new Uint8Array(new ArrayBuffer(bin.length));
  for (let i = 0; i < bin.length; i++) {
    out[i] = bin.charCodeAt(i);
  }
  return out;
}
